plenum/server/node.py

```diff
@@ -2354,15 +2354,42 @@ def lost_master_primary(self):
         self._schedule_view_change()
 
     def select_primaries(self, nodeReg: Dict[str, HA]=None):
+        primaries = set()
+        primary_rank = None
+        '''
+        Build a set of names of primaries, it is needed to avoid
+        duplicates of primary nodes for different replicas.
+        '''
+        for instance_id, replica in enumerate(self.replicas):
+            if replica.primaryName is not None:
+                name = replica.primaryName.split(":", 1)[0]
+                primaries.add(name)
+                '''
+                Remember the rank of primary of master instance, it is needed
+                for calculation of primaries for backup instances.
+                '''
+                if instance_id == 0:
+                    primary_rank = self.poolManager.get_rank_by_name(
+                        replica.primaryName.split(":", 1)[0], nodeReg)
```

Comment:
Because replica.primaryName in fact is not just node name, it has a form "node_name:instance_id". So it would be nice to rename replica.primaryName to replica.primaryInstanceName, I'll check how wide re-factoring it will cause.